tests/rptest/tests/full_disk_test.py

```diff
@@ -0,0 +1,265 @@
+                              retries=10) for _ in range(len(self.topics))
+            ]
+        return self.producers
+
+    def _setup(self) -> None:
+        self.logger.debug(f"brokers: {self.redpanda.brokers_list()}")
+        self.admin = Admin(self.redpanda)
+
+        # python kafka client doesn't seem to recover from no leader for partition,
+        # despite retries, when it encounters ECONNREFUSED. The sleep here works
+        # around this.
+        sleep(5)
+
+    def _send_all_topics(self, msg: str, expect_blocked=False):
+        """ Send `msg` to all topics, retrying a fixed number of times for
+        expected success or failure. """
+        futures = []
+        num_topics = len(self.topics)
+        pause: float = num_topics * 1.0 / MAX_MSG_PER_SEC
+        was_blocked = False
+        success = False
+        t0 = time()
+        i = 0
+        producers = self._get_producers(refresh=True)
+        for _ in range(MAX_MSG // num_topics):
+            sleep(pause)
+            try:
+                for j, topic in enumerate(self.topic_names):
+                    producer = producers[j]
+                    self.logger.debug(f"send msg {i},{j}, t={time()-t0:.2f}")
+                    i = i + 1
+                    f = producer.send(topic, msg)
+                    f.get(30)
+            except BrokerNotAvailableError as e1:
+                was_blocked = True
+                if expect_blocked:
+                    self.logger.info(f"Write rejected as expected: {e1}")
+                    break
+                else:
+                    # health monitor reports take time to propagate
+                    self.logger.debug("Still being blocked, retrying.")
+                    continue
+            except NotLeaderForPartitionError as e2:
+                self.logger.debug(f"Not leader for partition {e2}--retrying.")
+                continue
+            # no exception, success
+            if not expect_blocked:
+                success = True
+                break
+        if expect_blocked:
+            assert was_blocked, "Failed to block write for full disk"
+        else:
+            assert success, "Failed to produce"
+
+    # XXX see class comment      /|\
+    @cluster(num_nodes=3, log_allow_list=FDT_LOG_ALLOW_LIST)
+    def test_write_rejection(self):
+        """ Verify that we block external writers when free disk space on any
+        node is below the threshold. """
+
+        self._setup()
+
+        for i in range(LOOP_ITERATIONS):
+            self.logger.info(f"Iteration {i} of {LOOP_ITERATIONS}..")
+            # 1. Confirm we can produce successfully.
+            self._send_all_topics("Q: What do redpandas eat?")
+
+            # 2. Set threshold below current free space, assert we get rejected.
+            self.full_disk.trigger_low_space()
+            self._send_all_topics(
+                "A: I don't know, but they can do a lot of bites per second.",
+                expect_blocked=True)
```

Comment:
That is covered in local_monitor unit tests. Also note that if an alert is not raised, this test will fail, since the alert enums are what drive the degraded state.

Reply:
what i mean is that: how do we distinguish between the client produce failing because of out-of-space condition vs other cases?

Reply:
In general I dislike this test--it was too fussy to get working correctly with the kafka-python/librdkafka client. I sort of want to delete it, but it is passing and covering the error code. If it ever is flaky in CI I will probably delete it. That, or find a more direct client where I can easily see error propagation without weird side effects (i.e. producer stuck on not leader for partition until I create a new client) and retries under the covers.

Reply:
I did take this suggestion the first time and wrote another test for it. See the second test class in full_disk_test.py [here.](https://github.com/redpanda-data/redpanda/pull/4803/commits/2992c0b8b5a1cd0d0e86f74bca7e1f2e450eeaae#diff-bd71830f26776b7150ee955bfe36a3c3ab6e6e8a32b01cc9d5679224bcb22d1cR234-R265) It turned out nicer than the first one, to the point that I was saying "Noah was right" under my breath. 😆